ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -1167,12 +1219,18 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
   /**
    * We can have duplicate entries in the message entries since updates can happen to the same key. For example,
    * insert a key followed by a delete. This would create two entries in the journal or the index. A single findInfo
-   * could read both the entries. The findInfo should return as clean information as possible. This method removes
-   * the oldest duplicate in the list.
+   * could read both the entries. The findInfo should return as clean information as possible.
+   * <p/>
+   * This function choose a delete entry over all entries but chooses a put entry over a ttl update entry
    * @param messageEntries The message entry list where duplicates need to be removed
    */
   private void eliminateDuplicates(List<MessageInfo> messageEntries) {
-    Set<StoreKey> setToFindDuplicate = new HashSet<StoreKey>();
+    Set<StoreKey> setToFindDuplicate = new HashSet<>();
+    // first remove the ttl update MessageInfo of any key whose put MessageInfo is already present
+    messageEntries.removeIf(
+        messageInfo -> !messageInfo.isDeleted() && !setToFindDuplicate.add(messageInfo.getStoreKey()));
+    // next, remove any put/ttl update if a delete is present
+    setToFindDuplicate.clear();
     ListIterator<MessageInfo> messageEntriesIterator = messageEntries.listIterator(messageEntries.size());
     while (messageEntriesIterator.hasPrevious()) {
       MessageInfo messageInfo = messageEntriesIterator.previous();
```

Comment:
this code could probably be simplified as well...

Reply:
Done